hbase-mapreduce/src/main/java/org/apache/hadoop/hbase/mapreduce/replication/VerifyReplication.java

```diff
@@ -240,55 +262,77 @@ public void map(ImmutableBytesWritable row, final Result value, Context context)
-          Result replicatedResult = replicatedTable.get(new Get(row.getRow()));
-          Result.compareResults(sourceResult, replicatedResult, false);
-          if (!sourceResult.isEmpty()) {
-            context.getCounter(Counters.GOODROWS).increment(1);
-            if (verbose) {
-              LOG.info("Good row key (with recompare): " + delimiter
-                + Bytes.toStringBinary(row.getRow()) + delimiter);
-            }
-          }
-          return;
-        } catch (Exception e) {
-          LOG.error("recompare fail after sleep, rowkey=" + delimiter
-            + Bytes.toStringBinary(row.getRow()) + delimiter);
-        }
+    @SuppressWarnings("FutureReturnValueIgnored")
+    private void logFailRowAndIncreaseCounter(Context context, Counters counter, Result row,
+      Result replicatedRow) {
+      byte[] rowKey = getRow(row, replicatedRow);
+      if (sleepMsBeforeReCompare == 0) {
+        context.getCounter(counter).increment(1);
+        context.getCounter(Counters.BADROWS).increment(1);
+        LOG.error("{}, rowkey={}{}{}", counter, delimiter, Bytes.toStringBinary(rowKey), delimiter);
+        return;
       }
-      context.getCounter(counter).increment(1);
-      context.getCounter(Counters.BADROWS).increment(1);
-      LOG.error(counter.toString() + ", rowkey=" + delimiter + Bytes.toStringBinary(row.getRow())
-        + delimiter);
+
+      VerifyReplicationRecompareRunnable runnable = new VerifyReplicationRecompareRunnable(context,
+        row, replicatedRow, counter, delimiter, tableScan, sourceTable, replicatedTable,
+        reCompareTries, sleepMsBeforeReCompare, reCompareBackoffExponent, verbose);
+
+      if (reCompareExecutor == null) {
+        runnable.run();
+        return;
+      }
+
+      reCompareExecutor.submit(runnable);
     }
 
     @Override
     protected void cleanup(Context context) {
+      if (reCompareExecutor != null && !reCompareExecutor.isShutdown()) {
+        reCompareExecutor.shutdown();
+        try {
+          boolean terminated = reCompareExecutor.awaitTermination(1, TimeUnit.MINUTES);
+          if (!terminated) {
+            List<Runnable> queue = reCompareExecutor.shutdownNow();
+            for (Runnable runnable : queue) {
+              ((VerifyReplicationRecompareRunnable) runnable).fail();
+            }
+
+            terminated = reCompareExecutor.awaitTermination(1, TimeUnit.MINUTES);
+
+            if (!terminated) {
+              int activeCount = Math.max(1, reCompareExecutor.getActiveCount());
+              LOG.warn("Found {} possible recompares still running in the executable"
+                + " incrementing BADROWS and FAILED_RECOMPARE", activeCount);
+              context.getCounter(Counters.BADROWS).increment(activeCount);
+              context.getCounter(Counters.FAILED_RECOMPARE).increment(activeCount);
+            }
+          }
+        } catch (InterruptedException e) {
+          LOG.error("fail to await executor termination in cleanup", e);
```

Comment:
i think we need to at least increment BADROWS/FAILED_RECOMPARE here, or otherwise throw an exception. I think throwing an exception probably makes sense, since we have no idea what the state of pending recompares was

Reply:
opted for throwing b/c as you mentioned we don't really have any idea what's going on within the executor at the time